File: init_db.py
"""
init_db.py
==========
Initializes the local SQLite database used by the Quantified Self MCP
server (server.py), from a health-data export file.

Usage:
    python init_db.py path/to/health.csv
    python init_db.py path/to/export.xml            # Apple Health export
    python init_db.py path/to/health.csv --source csv

By default the source format is guessed from the file extension (--source
auto, the default: .xml -> apple-health, anything else -> this project's
own csv format below) — pass --source explicitly to override that guess.
See import_adapters.py to add support for another export format; adding
one there is all that's needed; nothing below has to change.

By default the database is created at data/health.db next to this script
if that's writable, or wherever the HEALTH_DB_PATH environment variable
points (the same variable server.py reads, so both agree on the location
automatically). If data/ next to this script isn't writable — the usual
case for a system-wide `pip install`, where this script lives inside
site-packages — it falls back to a per-user data directory instead (see
logic.default_data_dir). Pass --db-path to override any of this for a
single run.

Required CSV columns (header names are matched case-insensitively):
    date

The original columns (steps, sleep_hours, resting_heart_rate) and the
newer ones (weight_kg, workout_minutes, mood, water_ml) are all read only
if present in a given CSV's header — this is what makes the incremental
update described below work for any of them, not just the newer ones.

Dates should be YYYY-MM-DD; MM/DD/YYYY is also accepted. Numbers may
include "$" and "," (stripped automatically, kept for consistency with
the shared parsing helpers).

Re-running upserts by date, so it's safe to re-run as you add more days —
or to add a column later: a CSV with only date and weight_kg, say, updates
just that column and leaves steps/sleep_hours/etc. for that date
untouched, rather than blanking them out. Pass --replace to clear the
table first instead. Rows with a problem (bad date, non-numeric value,
etc.) are skipped with a warning rather than aborting the whole import —
the final line printed always tells you how many rows loaded vs. were
skipped. All of this applies equally to non-CSV sources (see
import_adapters.py); only the parsing of the source file itself differs.
"""

from __future__ import annotations

import argparse
import csv
import os
import sys
from datetime import datetime
from pathlib import Path

from import_adapters import ADAPTERS, RowError, detect_adapter
from logic import connect_writable, default_data_dir, ensure_schema, upsert_metrics, validate_metrics

BASE_DIR = Path(__file__).parent.resolve()
DATA_DIR = default_data_dir(BASE_DIR)

# Matches server.py's HEALTH_DB_PATH convention, so both halves of the
# project agree on where the database lives without extra configuration.
DEFAULT_DB_PATH = Path(os.environ.get("HEALTH_DB_PATH", DATA_DIR / "health.db")).expanduser()

# "date" is the only column a CSV header must contain. Everything else is
# read only if present in that particular header (see _read_csv) — this
# list is the original set of columns from the project's first release,
# kept separate from METRIC_COLUMNS_ADDED_LATER only for that historical
# reason, not because either group is more "required" than the other.
CORE_METRIC_COLUMNS = ["steps", "sleep_hours", "resting_heart_rate"]

# Columns added after the original release. Kept separate from
# CORE_METRIC_COLUMNS only to document that history; treated identically.
OPTIONAL_COLUMNS = ["weight_kg", "workout_minutes", "mood", "water_ml"]

DATE_FORMATS = ["%Y-%m-%d", "%m/%d/%Y"]


def _normalize_date(raw: str) -> str:
    raw = raw.strip()
    for fmt in DATE_FORMATS:
        try:
            return datetime.strptime(raw, fmt).date().isoformat()
        except ValueError:
            continue
    raise RowError(f"unrecognized date {raw!r} (use YYYY-MM-DD or MM/DD/YYYY)")


def _clean_number(raw: str) -> str:
    return raw.strip().replace("$", "").replace(",", "")


def _to_int(raw: str) -> int | None:
    raw = _clean_number(raw)
    if not raw:
        return None
    try:
        return int(float(raw))
    except ValueError as exc:
        raise RowError(f"expected a number, got {raw!r}") from exc


def _to_float(raw: str) -> float | None:
    raw = _clean_number(raw)
    if not raw:
        return None
    try:
        return float(raw)
    except ValueError as exc:
        raise RowError(f"expected a number, got {raw!r}") from exc


def _read_csv(csv_path: Path) -> tuple[list[dict[str, str]], list[str]]:
    """Read csv_path, matching column names case-insensitively.

    Returns (rows, present_columns) — the latter is whichever metric
    columns (core or added-later) actually appeared in this CSV's header,
    so the caller knows which ones to parse, validate, and upsert versus
    leave untouched. "date" is the only column required to be in the
    header at all; a CSV with just date + one metric column is valid.
    """
    with csv_path.open(newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        if reader.fieldnames is None:
            sys.exit(f"Error: {csv_path} appears to be empty.")
        header_map = {name.strip().lower(): name for name in reader.fieldnames}
        if "date" not in header_map:
            sys.exit(
                f"Error: {csv_path} is missing required column: date. "
                f"Found columns: {', '.join(reader.fieldnames)}"
            )
        present_columns = [c for c in ALL_METRIC_COLUMNS if c in header_map]
        wanted = {"date": header_map["date"], **{c: header_map[c] for c in present_columns}}
        rows = []
        for raw_row in reader:
            rows.append({canonical: raw_row.get(original) for canonical, original in wanted.items()})
    return rows, present_columns


ALL_METRIC_COLUMNS = CORE_METRIC_COLUMNS + OPTIONAL_COLUMNS

# Parser for each metric column's raw CSV string.
_METRIC_PARSERS = {
    "steps": _to_int,
    "sleep_hours": _to_float,
    "resting_heart_rate": _to_int,
    "weight_kg": _to_float,
    "workout_minutes": _to_int,
    "mood": _to_int,
    "water_ml": _to_int,
}


def _load_csv_rows(csv_path: Path) -> tuple[list[dict], list[str], int]:
    """This project's own CSV format, exactly as before this module split
    out other adapters — unchanged so existing tests (and existing CSV
    exports people already have from this project) keep working exactly
    as they did. Returns (parsed_rows, present_columns, skipped) to match
    the shape init_health_db needs regardless of which source it read.
    """
    raw_rows, present_columns = _read_csv(csv_path)

    parsed_rows, skipped = [], 0
    for i, row in enumerate(raw_rows, start=2):  # +2: header is line 1
        try:
            date_val = (row.get("date") or "").strip()
            if not date_val:
                raise RowError("missing date")
            parsed = {"date": _normalize_date(date_val)}
            for col in present_columns:
                parsed[col] = _METRIC_PARSERS[col](row.get(col) or "")
            try:
                validate_metrics({k: v for k, v in parsed.items() if k != "date"})
            except ValueError as exc:
                raise RowError(str(exc)) from exc
            parsed_rows.append(parsed)
        except RowError as exc:
            print(f"Skipping {csv_path} line {i}: {exc}", file=sys.stderr)
            skipped += 1
    return parsed_rows, present_columns, skipped


def init_health_db(source_path: Path, db_path: Path, replace: bool, source: str = "auto") -> None:
    """Import source_path into db_path.

    source picks which import_adapters.ADAPTERS entry reads source_path;
    "auto" (the default) guesses from the file extension via
    import_adapters.detect_adapter, "csv" is always this project's own
    format (handled directly here — see _load_csv_rows), and any other
    name must be a key in import_adapters.ADAPTERS.
    """
    adapter_name = detect_adapter(source_path) if source == "auto" else source

    if adapter_name == "csv":
        parsed_rows, present_columns, skipped = _load_csv_rows(source_path)
    else:
        if adapter_name not in ADAPTERS:
            sys.exit(f"Error: unknown import source {adapter_name!r}. Available: csv, {', '.join(ADAPTERS)}")
        adapted = ADAPTERS[adapter_name](source_path)
        parsed_rows, skipped = [], adapted.skipped
        for row in adapted.rows:
            try:
                validate_metrics({k: v for k, v in row.items() if k != "date"})
                parsed_rows.append(row)
            except ValueError as exc:
                print(f"Skipping {source_path} date {row.get('date')}: {exc}", file=sys.stderr)
                skipped += 1
        present_columns = adapted.present_columns

    conn = connect_writable(db_path)
    try:
        ensure_schema(conn)
        if replace:
            conn.execute("DELETE FROM daily_metrics")
        upsert_metrics(conn, parsed_rows)
    finally:
        conn.close()

    if present_columns:
        print(f"Loaded columns: {', '.join(present_columns)}")
    print(
        f"Health DB ready at {db_path}: {len(parsed_rows)} row(s) loaded, "
        f"{skipped} skipped. (source: {adapter_name})"
    )


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Load a health-data export into the Quantified Self health SQLite database."
    )
    parser.add_argument(
        "source_path",
        type=Path,
        help="Path to the source export file (this project's CSV, or e.g. Apple Health's export.xml).",
    )
    parser.add_argument(
        "--source",
        choices=["auto", "csv", *ADAPTERS],
        default="auto",
        help=(
            "Which import format source_path is. 'auto' (default) guesses "
            "from the file extension: .xml -> apple-health, anything else "
            "-> csv. See import_adapters.py for what each non-csv adapter "
            "supports."
        ),
    )
    parser.add_argument(
        "--replace",
        action="store_true",
        help="Clear existing rows in the table first, instead of appending/upserting.",
    )
    parser.add_argument(
        "--db-path",
        type=Path,
        default=DEFAULT_DB_PATH,
        help=(
            "Where to create/update the SQLite database. Defaults to the "
            "HEALTH_DB_PATH environment variable if set (same variable "
            "server.py reads), otherwise data/health.db next to this script "
            f"(currently: {DEFAULT_DB_PATH})."
        ),
    )
    args = parser.parse_args()

    if not args.source_path.exists():
        sys.exit(f"Error: source file not found at {args.source_path}")

    db_path = args.db_path.expanduser()
    db_path.parent.mkdir(parents=True, exist_ok=True)
    init_health_db(args.source_path, db_path, args.replace, args.source)


if __name__ == "__main__":
    main()

File: node/src/init-db.ts
#!/usr/bin/env node
import { existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import {
  connectWritable,
  defaultDataDir,
  ensureSchema,
  insertMeasurement,
  upsertMetrics,
  validateMetrics,
} from "./logic.js";
import { parseCsvFile, resolveHeader } from "./csv.js";
import { ADAPTERS, RowError, detectAdapter } from "./import-adapters.js";

// Base directory for the default data/ location: the current working
// directory the command is run from, not the script's own install
// location (which would resolve inside node_modules for an npm install —
// see logic.defaultDataDir's docstring for the full rationale).
const BASE_DIR = process.cwd();
const DATA_DIR = process.env.HEALTH_DB_PATH ? dirname(process.env.HEALTH_DB_PATH) : defaultDataDir(BASE_DIR);
const DEFAULT_DB_PATH = process.env.HEALTH_DB_PATH ?? resolve(DATA_DIR, "health.db");

const CORE_METRIC_COLUMNS = ["steps", "sleep_hours", "resting_heart_rate"];
const OPTIONAL_COLUMNS = ["weight_kg", "workout_minutes", "mood", "water_ml", "heart_rate", "hrv_ms"];
const ALL_METRIC_COLUMNS = [...CORE_METRIC_COLUMNS, ...OPTIONAL_COLUMNS];

function die(message: string): never {
  console.error(`Error: ${message}`);
  process.exit(1);
}

function normalizeDate(raw: string): string {
  const s = raw.trim();
  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) {
    if (Number.isNaN(new Date(s + "T00:00:00Z").getTime())) {
      throw new RowError(`unrecognized date '${raw}' (use YYYY-MM-DD or MM/DD/YYYY)`);
    }
    return s;
  }
  const m = /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/.exec(s);
  if (m) {
    const [, mm, dd, yyyy] = m;
    const iso = `${yyyy}-${mm.padStart(2, "0")}-${dd.padStart(2, "0")}`;
    if (Number.isNaN(new Date(iso + "T00:00:00Z").getTime())) {
      throw new RowError(`unrecognized date '${raw}' (use YYYY-MM-DD or MM/DD/YYYY)`);
    }
    return iso;
  }
  throw new RowError(`unrecognized date '${raw}' (use YYYY-MM-DD or MM/DD/YYYY)`);
}

function cleanNumber(raw: string): string {
  return raw.trim().replace(/[$,]/g, "");
}

function toInt(raw: string): number | null {
  const cleaned = cleanNumber(raw);
  if (!cleaned) return null;
  const n = Number(cleaned);
  if (!Number.isFinite(n)) throw new RowError(`expected a number, got '${raw}'`);
  return Math.trunc(n);
}

function toFloat(raw: string): number | null {
  const cleaned = cleanNumber(raw);
  if (!cleaned) return null;
  const n = Number(cleaned);
  if (!Number.isFinite(n)) throw new RowError(`expected a number, got '${raw}'`);
  return n;
}

const METRIC_PARSERS: Record<string, (raw: string) => number | null> = {
  steps: toInt,
  sleep_hours: toFloat,
  resting_heart_rate: toInt,
  weight_kg: toFloat,
  workout_minutes: toInt,
  mood: toInt,
  water_ml: toInt,
  heart_rate: toInt,
  hrv_ms: toFloat,
};

function loadCsvRows(
  csvPath: string,
  columnMap: Record<string, string>
): { parsedRows: Record<string, unknown>[]; presentColumns: string[]; skipped: number } {
  const text = readFileSync(csvPath, "utf-8").replace(/^\uFEFF/, "");
  const { header, rows } = parseCsvFile(text);
  if (header.length === 0) die(`${csvPath} appears to be empty.`);

  const dateCol = resolveHeader(header, "date", columnMap);
  if (!dateCol) {
    if (columnMap.date) {
      die(`--map date=${columnMap.date} but ${csvPath} has no such column. Found columns: ${header.join(", ")}`);
    }
    die(`${csvPath} is missing required column: date (use --map date=<your column name> if it's named differently). Found columns: ${header.join(", ")}`);
  }
  for (const [canonical, mapped] of Object.entries(columnMap)) {
    if (canonical !== "date" && !header.includes(mapped)) {
      die(`--map ${canonical}=${mapped} but ${csvPath} has no such column. Found columns: ${header.join(", ")}`);
    }
  }

  const presentColumns = ALL_METRIC_COLUMNS.filter((c) => resolveHeader(header, c, columnMap) !== null);
  const colMap: Record<string, string> = { date: dateCol };
  for (const c of presentColumns) colMap[c] = resolveHeader(header, c, columnMap)!;

  const parsedRows: Record<string, unknown>[] = [];
  let skipped = 0;
  rows.forEach((raw, i) => {
    const lineNo = i + 2; // +2: header is line 1
    try {
      const dateVal = (raw[colMap.date] ?? "").trim();
      if (!dateVal) throw new RowError("missing date");
      const parsed: Record<string, unknown> = { date: normalizeDate(dateVal) };
      for (const col of presentColumns) {
        parsed[col] = METRIC_PARSERS[col]((raw[colMap[col]] ?? "").trim());
      }
      try {
        validateMetrics(Object.fromEntries(Object.entries(parsed).filter(([k]) => k !== "date")) as Record<string, number | null>);
      } catch (exc) {
        throw new RowError((exc as Error).message);
      }
      parsedRows.push(parsed);
    } catch (exc) {
      if (exc instanceof RowError) {
        console.error(`Skipping ${csvPath} line ${lineNo}: ${exc.message}`);
        skipped++;
      } else {
        throw exc;
      }
    }
  });

  return { parsedRows, presentColumns, skipped };
}

async function initHealthDb(
  sourcePath: string,
  dbPath: string,
  replace: boolean,
  source: string,
  columnMap: Record<string, string>
) {
  const adapterName = source === "auto" ? detectAdapter(sourcePath) : source;
  let parsedRows: Record<string, unknown>[] = [];
  let presentColumns: string[] = [];
  let skipped = 0;
  let rawMeasurements: { timestamp: string; metric: string; value: number; unit?: string | null; source?: string | null }[] = [];

  if (adapterName === "csv") {
    const result = loadCsvRows(sourcePath, columnMap);
    parsedRows = result.parsedRows;
    presentColumns = result.presentColumns;
    skipped = result.skipped;
  } else {
    if (!(adapterName in ADAPTERS)) {
      die(`unknown import source '${adapterName}'. Available: csv, ${Object.keys(ADAPTERS).join(", ")}`);
    }
    const adapted = await ADAPTERS[adapterName](sourcePath);
    skipped = adapted.skipped;
    for (const row of adapted.rows) {
      try {
        validateMetrics(Object.fromEntries(Object.entries(row).filter(([k]) => k !== "date")) as Record<string, number | null>);
        parsedRows.push(row);
      } catch (exc) {
        console.error(`Skipping ${sourcePath} date ${row.date}: ${(exc as Error).message}`);
        skipped++;
      }
    }
    presentColumns = adapted.present_columns;
    rawMeasurements = adapted.raw_measurements;
  }

  const db = connectWritable(dbPath);
  try {
    ensureSchema(db);
    if (replace) db.exec("DELETE FROM daily_metrics");
    upsertMetrics(db, parsedRows);
    if (rawMeasurements.length > 0) {
      const importedAt = new Date().toISOString().slice(0, 19);
      for (const m of rawMeasurements) {
        insertMeasurement(db, {
          timestamp: m.timestamp,
          metric: m.metric,
          value: m.value,
          unit: m.unit,
          source: m.source,
          source_type: "wearable",
          importer: adapterName,
          imported_at: importedAt,
        });
      }
    }
  } finally {
    db.close();
  }

  if (presentColumns.length) console.log(`Loaded columns: ${presentColumns.join(", ")}`);
  if (rawMeasurements.length) console.log(`Loaded ${rawMeasurements.length} raw measurement(s) with source provenance.`);
  console.log(`Health DB ready at ${dbPath}: ${parsedRows.length} row(s) loaded, ${skipped} skipped. (source: ${adapterName})`);
}

function parseArgs(argv: string[]) {
  const args = { sourcePath: "", source: "auto", replace: false, dbPath: DEFAULT_DB_PATH, columnMap: {} as Record<string, string> };
  const positional: string[] = [];
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--source") args.source = argv[++i];
    else if (a === "--replace") args.replace = true;
    else if (a === "--db-path") args.dbPath = argv[++i];
    else if (a === "--map") {
      const item = argv[++i];
      const eq = item.indexOf("=");
      if (eq === -1) die(`--map expects COLUMN=HEADER, got '${item}'`);
      const canonical = item.slice(0, eq).trim().toLowerCase();
      const header = item.slice(eq + 1).trim();
      if (canonical !== "date" && !ALL_METRIC_COLUMNS.includes(canonical)) {
        die(`--map column '${canonical}' is not recognized. Valid: date, ${ALL_METRIC_COLUMNS.join(", ")}`);
      }
      if (!header) die(`--map '${item}' has an empty header`);
      args.columnMap[canonical] = header;
    } else if (a === "--help" || a === "-h") {
      printUsage();
      process.exit(0);
    } else positional.push(a);
  }
  if (positional.length === 0) {
    printUsage();
    process.exit(1);
  }
  args.sourcePath = positional[0];
  return args;
}

function printUsage() {
  console.error(
    "Usage: quantified-self-init-db <source-file> [--source auto|csv|apple-health|health-connect] " +
      "[--map COLUMN=HEADER]... [--replace] [--db-path PATH]"
  );
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (!existsSync(args.sourcePath)) die(`source file not found at ${args.sourcePath}`);
  mkdirSync(dirname(args.dbPath), { recursive: true });
  await initHealthDb(args.sourcePath, args.dbPath, args.replace, args.source, args.columnMap);
}

main().catch((err) => {
  console.error(err instanceof Error ? err.message : err);
  process.exit(1);
});
